test/ecto/changeset_test.exs

```diff
@@ -46,6 +46,28 @@ defmodule Ecto.ChangesetTest do
     assert changeset.valid?
   end
 
+  test "cast/3: optional field of wrong type is marked as invalid" do
+    params = %{"body" => :world}
+    struct = %Post{}
+
+    changeset = cast(struct, params, ~w(body))
+    assert changeset.changes == %{}
+    assert changeset.errors == [body: {"is invalid", [type: :string]}]
+    refute changeset.valid?
+  end
+
+  test "cast/3: required field (via validate_required/2) of wrong type is marked as invalid" do
```

Comment:
I'm curious if others think this is the right way to test this.  The setup uses both `cast` and `validate_required`, since that is the use case that spawned this change.  However, we can also test it in `validate_required` just by showing that pre-existing errors on a field prevent further errors.

Note that both tests I added are patterned after similar tests in `cast/4` that check for required vs. optional fields that have an invalid type.